learningequality/studio
# Fix star/bookmarking feature in Content Library

## Summary
### Description of the change(s) you made

- Added `ChannelListPagination` class for `ChannelViewSet`
- Added tests in `tests_channel.py` to test bookmarking feature
- Added `update_from_changes` method to `ChannelViewSet` class to override the one in `UpdateModelMixin` so that any user can bookmark any channel (the initial problem was that `update_from_changes` in `UpdateModelMixin` only worked with channels that a user could edit, not ones they could only view)


### Manual verification steps performed
1. Log in as a non-admin user w/view-only access to a Published Channel
2. Click on the "Content Library" tab
3. Bookmark the "Published Channel" by clicking on the star icon on its card
4. After a few seconds, click on the "Starred" filter checkbox on the left-hand side under "Formats" to filter all bookmarked channels
5. Check to see that there is 1 matching result
6. Unstar "Published Channel" by clicking on the star icon on the card
7. Uncheck the "Starred" filter. Wait a few seconds
8. Click on the "Starred" filter again
9. Check to see that there are "0 results found"

Note: There is a known bug (also see below) where quick clicking on the star icon and on the "Starred" filter will yield incorrect results, and the user will have to refresh the page in order for the accurate results to occur. If this occurs, please refresh the page to see that the correct results do show up after the refresh.

### Screenshots (if applicable)

![2021-04-09 22 12 52](https://user-images.githubusercontent.com/13563002/114259162-ee6e4600-9980-11eb-94f6-aba5b8d7fd8f.gif)


### Does this introduce any tech-debt items?

We will need to revisit pagination for channels.

___
## Reviewer guidance
### How can a reviewer test these changes?

Please manually test, and note that there is an edge case where if a user bookmarks and clicks on the "Starred" filter very quickly, it will not work (they will have to refresh the page).

Fixes #2933 

----

### Contributor's Checklist


Studio-specifc:

Testing:

___

### Reviewer's Checklist
#### This section is for reviewers to fill out.
